frontend/src/metabase/static-viz/components/LineAreaBarChart/utils/series.tsx

```diff
@@ -0,0 +1,174 @@
+    ? _.mapObject(settings.series_settings, value => {
+        return value.title;
+      })
+    : undefined;
+
+  let index = -1;
+  const legends = multipleSeries
+    .flatMap((questionSeries, seriesIndex) => {
+      return questionSeries.map(series => {
+        index++;
+
+        if (!hasTwoDimensions(series)) {
+          // One or zero dimensions
+
+          if (seriesIndex === 0 && series.name) {
+            return series.name;
+          }
+
+          const hasOneMetric = questionSeries.length === 1;
+          if (hasOneMetric) {
+            return series.cardName;
+          }
+
+          if (!isMultipleSeries) {
+            return series.column.display_name;
+          }
+
+          // is multiple series card
+          return `${series.cardName}: ${series.column.display_name}`;
+        } else {
+          // Two dimensions
+
+          const columnKey = formatStaticValue(series.breakoutValue, {
+            column: series.column,
+          });
+
+          if (!isMultipleSeries) {
+            return seriesTitles?.[keys[index]] ?? columnKey;
+          }
+
+          // is multiple series card
+          return `${series.cardName}: ${columnKey}`;
+        }
+      });
+    })
+    .filter(isNotNull);
+
+  index = -1;
+  return multipleSeries.map(questionSeries =>
+    questionSeries.map(series => {
+      index++;
+
+      return merge(series, {
+        name: legends[index],
+      });
+    }),
+  );
+}
+
+function getSeriesKeys(
+  multipleSeries: (SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions)[][],
+  isMultipleSeries: boolean,
+) {
+  return multipleSeries
+    .flatMap((questionSeries, seriesIndex) => {
+      return questionSeries.map(series => {
+        if (!hasTwoDimensions(series)) {
+          // One or zero dimensions
+          if (seriesIndex === 0) {
+            return series.column.name;
+          }
+
+          const hasOneMetric = questionSeries.length === 1;
+          if (!isMultipleSeries || hasOneMetric) {
+            return series.cardName;
+          }
+
+          // is multiple series card
+          return `${series.cardName}: ${series.column.display_name}`;
+        } else {
+          // Two dimension
+          const columnKey = formatStaticValue(series.breakoutValue, {
+            column: series.column,
+          });
+
+          if (!isMultipleSeries) {
+            return columnKey;
+          }
+
+          // is multiple series card
+          return `${series.cardName}: ${columnKey}`;
+        }
+      });
+    })
+    .filter(isNotNull);
+}
+
+function hasTwoDimensions(
+  series: SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions,
+): series is SeriesWithTwoDimensions {
+  return "breakoutValue" in series;
+}
+
+export function removeNoneSeriesFields(
+  series: (SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions)[][],
+): Series[] {
+  return series
+    .flat()
+    .map(
+      series => _.omit(series, "cardName", "column", "breakoutValue") as Series,
```

Comment:
This is kind of optional, but would make sure `series` passed to the rest of the combo chart has the same exact shape as before (has no `cardName`, `column`, or `breakoutValue` properties)